### DISCUSSION TITLE
Saving the calibration file

### DISCUSSION THREAD

**Discussion by Rtms1984** [openPupil/Open-PupilEXT#6] [Q&A] [answered]
Hi, for some reasons I can't manage to save the calibration .xml file in stereo camera mode. It is working fine in single camera mode. I'm using the Windows release.

**Comment by BZandi** [comment]
What happens if you press the save button directly after the successfully performed stereo calibration ?

**Reply by Rtms1984** [reply]
I choose the directory and put a name for the file but it saves nothing.

**Comment by BZandi** [comment]
I could not reproduce this issue. Today, I tested it on a Windows machine with the PupilEXT 0.1.1 release, and everything is working. Maybe you use an invalid name. See the attached video, which I recorded during today's calibration.

https://user-images.githubusercontent.com/62992112/128506317-992fa529-751d-41ec-b706-5319033272b7.mp4

**Reply by BZandi** [reply]
The last status of PupilEXT and all settings are always saved in a preference *.ini file. Therefore you can try to reset the PupilEXT settings to default. Maybe there is an issue with that on your machine. For this, plug out both cameras from your computer and open PupilEXT. 

**Step 1**: Go to "Windows" and press "Reset Interface". 
**Step 2**: Delete the preference *.ini file from PupilEXT on your local machine. To locate
the preference file, go to the "about window" of PupilEXT, and your application settings path should be stated in the "getting wizard" window in the last line. Locate the file on your machine and delete it. PupilEXT should be closed when you delete the file. This should push everything to default.

**Reply by BZandi** [reply]
**Update**
I saw that the location of the preference file is not stated under "About" in the windows release. You can find the file under the following path: C:\Users\YourUserName\AppData\Roaming\FGLT\

Delete everything in this folder to reset PupilEXT.

**Reply by Rtms1984** [reply]
I did the same. Still doesn't work. I have the name of the files I tried to save in the Quick access under Recent files. But when I click on them, it says that windows cannot find them.

**Reply by BZandi** [reply]
It must be something specific with your PC settings, like permission to wright or similar. Try to save the file directly on your Desktop.

**Additional solution:**
When you close PupilEXT after successful calibration, the last calibration file is also always saved automatically (for backup reasons) under the following path: C:\Users\ ...YourUserName... \AppData\Roaming\FGLT\PupilEXT

Locate this folder and check if you see the calibration file there. The folder "AppData\" is a hidden folder. You will not see it under recent files. So you need to turn on "show hidden folder" in windows and locate this path manually. See here how to view hidden files:
https://support.microsoft.com/en-us/windows/view-hidden-files-and-folders-in-windows-10-97fbc472-c603-9d90-91d0-1166d1d9f4b5

**Reply by Rtms1984** [reply]
There is no file there also. If the problem is with the permission to write how come I can save the calibration files in single camera mode?

**Reply by BZandi** [reply]
> There is no file there also. If the problem is with the permission to write how come I can save the calibration files in single camera mode?

True, that's strange. As I cannot replicate this issue, it is hard to debug. Can you please take a screenshot to see which files are stored in the following folders:

C:\Users\ ...YourUserName... \AppData\Roaming\FGLT\PupilEXT\
and
C:\Users\ ...YourUserName... \AppData\Roaming\FGLT\

Do you have a space in the user name of your windows, like "PC John" or similar? 

@lodemo do you have an idea what could cause this issue?

**Reply by Rtms1984** [reply]
PupilEXT is empty and FGLT has this empty folder + PupilEXT.ini .
Yes there is a space in the user name.

**Reply by BZandi** [reply]
I think the space in your username could be the problem. Try to remove it. It is recommended to follow a tutorial for changing the username. Spaces should be avoided when setting a username. It causes issues. See this thread for example:

https://github.com/conda/conda/issues/8725

**Comment by BZandi** [comment, marked as answer]
Think I located the source of the problem. In the file ``PupilEXT/src/subwindows/stereoCameraCalibrationView.cpp`` (line 228 to 243), the calibration file will be saved with the following method:

```c++
void StereoCameraCalibrationView::onSaveClick() {
    QString filename = QFileDialog::getSaveFileName(this, tr("Save Calibration File"), "", tr("XML files (*.xml)"));

    if(!filename.isEmpty()) {
        filename.replace(" ", "");

        QFileInfo fileInfo(filename);
        // check if filename has extension, else append xml to it
        if(fileInfo.suffix().isEmpty()) {
            filename = filename + ".xml";
        }

        std::cout<<"Saving calibration file to settings directory: "<< filename.toStdString() <<std::endl;
        calibrationWorker->saveToFile(filename.toStdString().c_str());
    }
}
```
After the if condition with the code `filename.replace(" ", "");` all spaces will be removed from the selected path in the dialog window. For example, if you select the Desktop and your username has a space, the file will not be saved correctly, as the username is part of the path. This issue does not happen with the single camera calibration, as the code snippet for removing the spaces is not applied there. See the file`PupilEXT/src/subwindows/singleCameraCalibrationView.cpp ` in line 214 to 227.

As you have a space in your Windows username you will run into this when you save a stereo calibration file. This will be solved in the next release.
 
As a **workaround**, you can try to save the stereo calibration file directly into the `C:\` drive or another disc without a space or the username in the save path. Please, let me know if this workaround was successful for you.

**Reply by Rtms1984** [reply]
It worked. Thanks!

**Reply by QuinzyW** [reply]
thanks，I did really meet the same question.

**Comment by QuinzyW** [comment]
I use Win10 and can't save the `calibration XML file`. In the end, I find the PupilEXT should `Run as an administrator` to get the right to write `XML file` on `Desktop` or `C Disk`.
